pySDC/projects/DAE/problems/DiscontinuousTestDAE.py

```diff
@@ -0,0 +1,184 @@
+        t : float
+            Current time of the numerical solution.
+
+        Returns
+        -------
+        f : dtype_f
+            The right-hand side of f (contains two components).
+        """
+
+        y, z = u[0], u[1]
+        dy = du[0]
+
+        t_switch = np.inf if self.t_switch is None else self.t_switch
+
+        h = 2 * y - 100
+        f = self.dtype_f(self.init)
+
+        f_before_event = (
+            dy - z,
+            y**2 - z**2 - 1,
+        )
+
+        f_after_event = (
+            dy,
+            y**2 - z**2 - 1,
+        )
+
+        if h >= 0 or t >= t_switch:
+            f[:] = f_after_event
+        else:
+            f[:] = f_before_event
+        return f
+
+    def u_exact(self, t, u_init=None, t_init=None):
+        r"""
+        Routine for the exact solution at time :math:`t \leq 1`. For this problem, the exact
+        solution is piecewise.
+
+        Parameters
+        ----------
+        t : float
+            Time of the exact solution.
+
+        Returns
+        -------
+        me : dtype_u
+            Exact solution.
+        """
+
+        assert t >= 1, 'ERROR: u_exact only valid for t>=1'
+
+        me = self.dtype_u(self.init)
+        if t <= self.t_switch_exact:
+            me[:] = (np.cosh(t), np.sinh(t))
+        else:
+            me[:] = (np.cosh(self.t_switch_exact), np.sinh(self.t_switch_exact))
+        return me
+
+    def get_switching_info(self, u, t):
+        r"""
+        Provides information about the state function of the problem. When the state function changes its sign,
+        typically an event occurs. So the check for an event should be done in the way that the state function
+        is checked for a sign change. If this is the case, the intermediate value theorem states a root in this
+        step.
+
+        The state function for this problem is given by
+
+        .. math::
+           h(y(t)) = 2 y(t) - 100.
+
+        Parameters
+        ----------
+        u : dtype_u
+            Current values of the numerical solution at time :math:`t`.
+        t : float
+            Current time of the numerical solution.
+
+        Returns
+        -------
+        switch_detected : bool
+            Indicates whether a discrete event is found or not.
+        m_guess : int
+            The index before the sign changes.
+        state_function : list
+            Defines the values of the state function at collocation nodes where it changes the sign.
+        """
+
+        switch_detected = False
+        m_guess = -100
+
+        for m in range(1, len(u)):
+            h_prev_node = 2 * u[m - 1][0] - 100
+            h_curr_node = 2 * u[m][0] - 100
+            if h_prev_node < 0 and h_curr_node >= 0:
+                switch_detected = True
+                m_guess = m - 1
+                break
+
+        state_function = [2 * u[m][0] - 100 for m in range(len(u))]
+        return switch_detected, m_guess, state_function
+
+    def count_switches(self):
```

Comment:
Originally, I added this method for `battery_n_capacitors` with more than one capacitor s.t. the switching between the states is done correctly in `eval_f` and `solve_sytem`, because switching there bases (still) on the number of events found. Probably counting the events can also be done using you preferred `WorkCounter`'s, so I'll see how can I change that, since these method is (completely) unnecessary for all other problem classes containing events.